### ISSUE TITLE
Let's discuss sim2sim together #23

### ISSUE THREAD

**Issue by SYK985** [OpenRobotLab/HoST#2]
Thank you for your excellent work. I am trying deployment, but work not so well, code as follow.If the result is better and more convenient, please open source it. Thanks.

import time

import mujoco.viewer
import mujoco
import numpy as np
import torch
import yaml
import os
from pathlib import Path
from collections import deque
import matplotlib.pyplot as plt
import csv
import pickle

def get_gravity_orientation(quaternion):
    qw = quaternion[0]
    qx = quaternion[1]
    qy = quaternion[2]
    qz = quaternion[3]

    gravity_orientation = np.zeros(3)

    gravity_orientation[0] = 2 * (-qz * qx + qw * qy)
    gravity_orientation[1] = -2 * (qz * qy + qw * qx)
    gravity_orientation[2] = 1 - 2 * (qw * qw + qz * qz)

    return gravity_orientation

def pd_control(target_q, q, kp, target_dq, dq, kd):
    """Calculates torques from position commands"""
    return (target_q - q) * kp + (target_dq - dq) * kd

if __name__ == "__main__":
    # get config file name from command line
    import argparse

    current_path = os.path.dirname(__file__)

    config_path =os.path.join(current_path, "configs/g1.yaml")
    print("config_path: ", config_path)
    
    with open(config_path, "r") as f:
        config = yaml.load(f, Loader=yaml.FullLoader)

        policy_path = os.path.join(current_path, "../models/g1", config["policy_name"])
        xml_path    = os.path.join(current_path, "../../robots/g1", config["xml_name"])

        print("policy_path: ", policy_path)
        print("xml_path   : ", xml_path)

        simulation_duration = config["simulation_duration"]
        simulation_dt = config["simulation_dt"]
        control_decimation = config["control_decimation"]

        kps = np.array(config["kps"], dtype=np.float32)
        kds = np.array(config["kds"], dtype=np.float32)

        default_angles = np.array(config["default_angles"], dtype=np.float32)
        # print("****** default_angles ******", default_angles)

        ang_vel_scale = config["ang_vel_scale"]
        dof_pos_scale = config["dof_pos_scale"]
        dof_vel_scale = config["dof_vel_scale"]
        action_scale = config["action_scale"]
        cmd_scale = np.array(config["cmd_scale"], dtype=np.float32)

        num_actions = config["num_actions"]
        num_obs = config["num_obs"]

        cmd = np.array(config["cmd_init"], dtype=np.float32)

        num_mass_params_tensor = config["num_mass_params_tensor"]
        num_friction_coeffs_tensor = config["num_friction_coeffs_tensor"]
        num_lin_vel = config["num_lin_vel"]

        policy_mode = config["policy_mode"]

    # define context variables
    action = np.zeros(num_actions, dtype=np.float32)
    mass_params_tensor = np.zeros(num_mass_params_tensor, dtype=np.float32)
    friction_coeffs_tensor = 1.0*np.ones(num_friction_coeffs_tensor, dtype=np.float32)
    friction_coeffs_tensor = np.zeros(num_friction_coeffs_tensor, dtype=np.float32)

    lin_vel = np.zeros(num_lin_vel, dtype=np.float32)
    motro_stength_0 = np.ones(num_actions, dtype=np.float32)
    motro_stength_1 = np.ones(num_actions, dtype=np.float32)
    target_dof_pos = default_angles.copy()
    obs = np.zeros(num_obs, dtype=np.float32)

    counter = 0
    delay_buffer = np.zeros((5, 1, 23))

    # Load robot model
    m = mujoco.MjModel.from_xml_path(xml_path)
    d = mujoco.MjData(m)
    m.opt.timestep = simulation_dt

    # load policy
    policy = torch.jit.load(policy_path)

    if policy_mode =="unitree":
        obs_buf = np.zeros(num_obs, dtype=np.float32)
    elif  policy_mode == "host":
        obs_buf = np.zeros(num_obs * 6, dtype=np.float32)

    with mujoco.viewer.launch_passive(m, d) as viewer:
        
        # Close the viewer automatically after simulation_duration wall-seconds.
        start = time.time()
        while viewer.is_running() and time.time() - start < simulation_duration:

            step_start = time.time()
            tau = pd_control(target_dof_pos, d.qpos[7:], kps, np.zeros_like(kds), d.qvel[6:], kds)
            d.ctrl[:] = tau

            # mj_step can be replaced with code that also evaluates
            # a policy and applies a control signal before stepping the physics.
            mujoco.mj_step(m, d)

            counter += 1
            if counter % control_decimation == 0:
                # Apply control signal here.

                # create observation
                qj = d.qpos[7:]      # 关节角度
                dqj = d.qvel[6:]     # 关节速度
                quat = d.qpos[3:7]   # 四元数

                # omega = d.qvel[3:6]  # 角速度w
                ang_vel = d.qvel[3:6]
                qj = qj * dof_pos_scale
                dqj = dqj * dof_vel_scale
                gravity_orientation = get_gravity_orientation(quat)
                ang_vel = ang_vel * ang_vel_scale

                if policy_mode == "host":
                    current_obs = np.concatenate((
                                            ang_vel * 0.25,
                                            gravity_orientation,
                                            qj,
                                            dqj,
                                            action,
                                            np.array([0.3]) 
                                        ), axis=-1)

                    if counter / control_decimation <= 30.0:
                        current_obs = current_obs * 0

                    obs_buf = np.concatenate((current_obs, obs_buf[ 76:76*6]), axis=-1, dtype=np.float32)

                obs_tensor = torch.from_numpy(obs_buf).unsqueeze(0)

                action = policy(obs_tensor).detach().numpy().squeeze()

                if counter / control_decimation <= 30.0:
                    action = np.zeros(num_actions, dtype=np.float32)

                delay_idx = 2
                action_scaled = action * 1.0
                action_scaled_expanded = np.expand_dims(np.expand_dims(action_scaled, axis=0), axis=1)  # [1,1,23]
                delay_buffer = np.concatenate((delay_buffer[1:], action_scaled_expanded), axis=0)

                target_dof_pos = default_angles + delay_buffer[delay_idx, 0, :]

            # Pick up changes to the physics state, apply perturbations, update options from GUI.
            viewer.sync()

            # Rudimentary time keeping, will drift relative to wall clock.
            time_until_next_step = m.opt.timestep - (time.time() - step_start)
            if time_until_next_step > 0:
                time.sleep(time_until_next_step)

**Comment by TaoHuang13** [comment]
Hi. Please check the action space here. We use current pos + action as the PD target instead of default pos + action.

**Comment by speedronaldo666** [comment]
Thank you, it has been resolved.

**Comment by love1994love** [comment]
@speedronaldo666 Does it work well after the modification? I also deployed it on mujoco but found that the robot could not stand up.

**Comment by a123456aaq** [comment]
> [@speedronaldo666](https://github.com/speedronaldo666) Does it work well after the modification? I also deployed it on mujoco but found that the robot could not stand up.

Can we communicate together? I am currently deploying to mujoco, but I am also standing up and shaking wildly

**Comment by dbdxnuliba** [comment]
@TaoHuang13 ，could  you please provide sim2sim code about host ,many thanks

**Comment by yyl957491** [comment]
@speedronaldo666  Hello, may I ask if you have solved the sim2sim issue? I have tried to deploy it in Mujoco, but the result was not good. The robot couldn't stand up. Could we discuss it together? thank you

**Comment by GoGoGo13579** [comment]
> > [@speedronaldo666](https://github.com/speedronaldo666) Does it work well after the modification? I also deployed it on mujoco but found that the robot could not stand up.
> 
> Can we communicate together? I am currently deploying to mujoco, but I am also standing up and shaking wildly

I also encountered the issue of severe shaking when standing up in MuJoCo. Is there any way to solve it?

**Comment by ThelilinNB** [comment]
> > > [@speedronaldo666](https://github.com/speedronaldo666) Does it work well after the modification? I also deployed it on mujoco but found that the robot could not stand up.
> > 
> > 
> > Can we communicate together? I am currently deploying to mujoco, but I am also standing up and shaking wildly
> 
> I also encountered the issue of severe shaking when standing up in MuJoCo. Is there any way to solve it?

hi, may i ask if you have solved it?i couldn't get up when deploying in mujoco

**Comment by cj666201** [comment]
@TaoHuang13 ,could you please provide sim2sim code about host ,many thanks
